client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/OperationsFactory.java

```diff
@@ -75,7 +79,13 @@ public OperationsFactory(ChannelFactory channelFactory, String cacheName, boolea
             ? channelFactory.createTopologyId(cacheNameBytes)
             : new AtomicInteger(-1);
       this.forceReturnValue = forceReturnValue;
-      this.codec = codec;
+
+      Codec negotiatedCodec = (channelFactory != null) ? channelFactory.getNegotiatedCodec() : null;
+      this.codec = (negotiatedCodec != null) ? negotiatedCodec : codec;
+
+      // in order to support the handshake with 3.x clients
+      this.safePingVersionCodec = (codec instanceof Codec40) ? new Codec31() : codec;
```

Comment:
so we dont need to rely on the instanceof:

```suggestion
      this.safePingVersionCodec = codec.supportsTracing() ? ProtocolVersion.getWithoutTracing() : codec;
```

Reply:
I would avoid a for loop to choose a version that we know already is the 3.1